swarm/src/lib.rs

```diff
@@ -1731,3 +1777,134 @@ mod tests {
         }))
     }
 }
+
+pub mod dial_opts {
+    use libp2p_core::{Multiaddr, PeerId};
+
+    #[derive(Debug)]
+    pub enum DialOpts {
+        // TODO:  How about folding these structs into the enum variants?
+        WithPeerId(WithPeerId),
+        WithPeerIdWithAddresses(WithPeerIdWithAddresses),
+        WithoutPeerIdWithAddress(WithoutPeerIdWithAddress),
+    }
+
+    impl DialOpts {
```

Comment:
```suggestion
    enum Opts {
        // TODO:  How about folding these structs into the enum variants?
        WithPeerId(WithPeerId),
        WithPeerIdWithAddresses(WithPeerIdWithAddresses),
        WithoutPeerIdWithAddress(WithoutPeerIdWithAddress),
    }
    
    pub struct DialOpts {
        opts: Opts
    }

    impl DialOpts {
```
Consider wrapping the enum in a struct to emphasize that the user should not try to build the variants directly, but instead use the methods.

Reply:
Given the structs `WithPeerId`, `WithPeerIdWithAddresses` and `WithoutPeerIdWithAddress` can not be constructed outside of `dial_opts`, the user can not build `DialOpts` themselves. Still I understand that this is not intuitive and that it would be nice to make it clear that the user should go through the methods on `DialOpts`.

Is wrapping an enum into a struct an idiomatic way in Rust to signal that something should not be constructed directly? I don't recall seeing this in the past. Any prominent examples? I would have just relied on a doc comment on `DialOpts`.

Reply:
> Is wrapping an enum into a struct an idiomatic way in Rust to signal that something should not be constructed directly? I don't recall seeing this in the past.

I would argue it the other way around: I feel like it is very rare and not rust-idiomatic that an enum is **not** constructed directly. Of course a user would notice that it is not possible once they try to build the inner structs, but it still may cause some confusion. With a struct, the API would make it clearer that it should be constructed via methods, especially when looking at it on docs.rs.

> I don't recall seeing this in the past. Any prominent examples?

I feel like this is just following the general api design pattern of wrapping internal/ inner types in a new struct to hide implementation details. Imo this always makes sense when the inner type is not relevant for the user and should never be constructed directly by them, which is the case here.

---
Alternatively, if you decide to fold the structs into the enum variants as you wrote in your TODO in l.1786:
> `// TODO:  How about folding these structs into the enum variants`

the user would be able to build the enum types directly and the builder pattern would not be needed at all.

Reply:
Addressed in https://github.com/libp2p/rust-libp2p/pull/2317/commits/ad083e933a567f1f498e590f22a4022f0d09cfee.